src/services/spr.ts

```diff
@@ -0,0 +1,263 @@
+	if (!accountData) {
+		response.sendStatus(400);
+		return;
+	}
+
+	if (!accountData.friendCode || accountData.accessLevel < 0) {
+		response.sendStatus(400);
+		return;
+	}
+
+	const sprMetadataBuffer: Buffer | undefined = request.files['spr-meta'];
+	if (typeof sprMetadataBuffer === 'undefined') {
+		response.sendStatus(400);
+		return;
+	}
+
+	let sprSlots: SPRSlot[] = [];
+
+	// * Check spr-meta metadata headers
+	const sprMetadata = sprMetadataBuffer.toString();
+	const metadataHeaders = sprMetadata.split('\r\n'); // * Split header lines
+	if (metadataHeaders.length < 1) {
+		response.sendStatus(400);
+		return;
+	}
+
+	for (let i = 0; i < metadataHeaders.length; i++) {
+		const metadataHeader = metadataHeaders[i];
+		const field = metadataHeader.split(': '); // * Split header and value
+		if (field.length !== 2) {
+			response.sendStatus(400);
+			return;
+		}
+
+		// * Since the headers will always use the same pattern (first the slotsize, then the metadata for each slot),
+		// * we can guarantee that i must match with the slot we are looking at except for 0, which will be the slotsize
+		if (i === 0) {
+			if (field[0] !== 'slotsize') {
+				response.sendStatus(400);
+				return;
+			}
+
+			// * Validate slotsize
+			let slotsize: number;
+			try {
+				slotsize = parseInt(field[1]);
+			} catch {
+				response.sendStatus(400);
+				return;
+			}
+
+			// * We don't count the slotsize header itself in the slot count
+			if (slotsize !== (metadataHeaders.length - 1)) {
+				response.sendStatus(400);
+				return;
+			}
+
+			continue;
+		}
+
+		const metadata = field[1].split(','); // * Split the value to get the metadata
+		if (metadata.length !== 3) {
+			response.sendStatus(400);
+			return;
+		}
+
+		let sendMode: SendMode;
+		let gameID: number;
+		let size: number;
+		try {
+			sendMode = parseInt(metadata[0]);
+			gameID = parseInt(metadata[1], 16);
+			size = parseInt(metadata[2]);
+		} catch {
+			response.sendStatus(400);
+			return;
+		}
+
+		let data: Buffer = Buffer.alloc(0);
+		if (size > 0 && sendMode !== SendMode.RecvOnly) {
+			const slot = i.toString().padStart(2, '0');
+			const slotData: Buffer | undefined = request.files['spr-slot' + slot];
+			if (typeof slotData === 'undefined') {
+				response.sendStatus(400);
+				return;
+			}
+
+			if (slotData.length !== size) {
+				response.sendStatus(400);
+				return;
+			}
+
+			// * Integrity checks for slot data. Every StreetPass message sent over relays has the following header:
+			// * uint32 magic 0x6161
+			// * uint32 size
+			// * uint32 gameID
+			// * uint32 unknown1
+			// * uint32 unknown2
+			// * This is then followed by a CecMessageHeader (see https://github.com/NarcolepticK/CECDocs/blob/master/Structs/CecMessageHeader.md)
+
+			// * Check that we at least have enough size for the StreetPass header
+			if (slotData.length < 0x12) {
+				response.sendStatus(400);
+				return;
+			}
+
+			if (slotData.readUInt32LE(0) !== 0x6161) {
```

Comment:
Reading from buffers defaults to an offset of 0. You don't have to set it if you don't want to